[FX][EZ] Fix bug where copying node made non-unique name

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45227 [FX][WIP] Mutable Graph APIs
* #45245 [FX][WIP] Make return a node
* #44973 [FX] Lint pass for Graphs
* **#45311 [FX][EZ] Fix bug where copying node made non-unique name**

Copying a node with name e.g. `add_1` would go into `Graph._name` with `target='add_1'`. Since `add_1` was not found in `Graph._used_names` (because _used_names maps from base name to unique index, not full name), `add_1` would get assigned as the copied Node's name, potentially causing duplicate names

Differential Revision: [D23917864](https://our.internmc.facebook.com/intern/diff/D23917864)

Commit:
[FX][EZ] Fix bug where copying node made non-unique name

[ghstack-poisoned]

dzhulgakov commented on torch/fx/graph.py:
```diff
@@ -137,7 +137,7 @@ def node_copy(self, node: Node, arg_transform: Callable[[Node], Argument] = lamb
             name = node.name
         else:
-            name = self._name(node.name)
+            name = self._name(node.target)
```
what if my names are semantically meaningful (like some pass edited them) and I don't want to completely loose them?

I think it's better to actually track both versions in `_used_names`. E.g. you can add both with and without suffix to the dictionary (`{'add': 3, `add_1`: 0, `add_2`: 0, `add_3`: 0}`) - in the worst case it'd produce `add_1_1_!`.

Commit:
Update on "[FX][EZ] Fix bug where copying node made non-unique name"


Copying a node with name e.g. `add_1` would go into `Graph._name` with `target='add_1'`. Since `add_1` was not found in `Graph._used_names` (because _used_names maps from base name to unique index, not full name), `add_1` would get assigned as the copied Node's name, potentially causing duplicate names

Differential Revision: [D23917864](https://our.internmc.facebook.com/intern/diff/D23917864)

[ghstack-poisoned]

dzhulgakov commented on torch/fx/graph.py:
```diff
@@ -137,7 +137,15 @@ def node_copy(self, node: Node, arg_transform: Callable[[Node], Argument] = lamb
+                try:
+                    int(maybe_idx)
+                    sanitized_name = base
+                except:
```
make it `except ValueError`:

dzhulgakov commented on torch/fx/graph.py:
```diff
@@ -137,7 +137,15 @@ def node_copy(self, node: Node, arg_transform: Callable[[Node], Argument] = lamb
+                    sanitized_name = base
+                except:
+                    pass
+            name = self._name(sanitized_name)
```
I don't think it's fully sufficient for crazy cases when user passes `add_1_1` first (it's added without the index) and then they pass `add_1` (or `add` twice). You should tests the names with indices in the used dict too (inside `_name`), just to be safe.

dzhulgakov replied:
Still, my comment stands: you need to add both `self._used_names[op]` and `send._used_names[f'{op}_{i}']` inside `_register_name_used` to avoid (unlikely) situation when they first add `foo_1_1` and then `foo_1`

Commit:
Update on "[FX][EZ] Fix bug where copying node made non-unique name"


Copying a node with name e.g. `add_1` would go into `Graph._name` with `target='add_1'`. Since `add_1` was not found in `Graph._used_names` (because _used_names maps from base name to unique index, not full name), `add_1` would get assigned as the copied Node's name, potentially causing duplicate names

Differential Revision: [D23917864](https://our.internmc.facebook.com/intern/diff/D23917864)

[ghstack-poisoned]

Commit:
Update on "[FX][EZ] Fix bug where copying node made non-unique name"


Copying a node with name e.g. `add_1` would go into `Graph._name` with `target='add_1'`. Since `add_1` was not found in `Graph._used_names` (because _used_names maps from base name to unique index, not full name), `add_1` would get assigned as the copied Node's name, potentially causing duplicate names

Differential Revision: [D23917864](https://our.internmc.facebook.com/intern/diff/D23917864)

[ghstack-poisoned]

Commit:
Update on "[FX][EZ] Fix bug where copying node made non-unique name"


Copying a node with name e.g. `add_1` would go into `Graph._name` with `target='add_1'`. Since `add_1` was not found in `Graph._used_names` (because _used_names maps from base name to unique index, not full name), `add_1` would get assigned as the copied Node's name, potentially causing duplicate names

Differential Revision: [D23917864](https://our.internmc.facebook.com/intern/diff/D23917864)

[ghstack-poisoned]

jamesr66a commented on torch/quantization/fx/quantize.py:
```diff
@@ -108,10 +108,10 @@ def graph_module_from_producer_nodes(root, producer_nodes):
-        env[producer_node.name] = graph.node_copy(producer_node, load_arg)
-    graph.output(load_arg(producer_nodes[-1].name))
+        env[producer_node] = graph.node_copy(producer_node, load_arg)
+    graph.output(load_arg(producer_nodes[-1]))
```
@jerryzh168 you shouldn't be using the name of nodes to find correspondences between copies of the nodes in different graphs. The names are not guaranteed to be the same after copying. Please check the other places in the code where names are being used in map keys

jerryzh168 replied:
the map is between name in old graph to name of the corresponding node in new graph. I don't think I'm assuming the name of copied node is the same as the original node. Could you elaborate on why that matters?
